vfio_user/src/lib.rs

```diff
@@ -476,12 +477,23 @@ impl Client {
         let num_regions = reply.num_regions;
         let mut regions = Vec::new();
         for index in 0..num_regions {
-            let (region_info, fd) = self.get_region_info(index, None)?;
+            let (mut region_info, mut fd, mut sparse_areas) = self.get_region_info(index, None)?;
+            if region_info.argsz > std::mem::size_of::<vfio_region_info>() as u32 {
+                // Note: workaround for https://github.com/rust-lang/rust/issues/71126 to compile with rust toolchian 1.56
+                let (_region_info, _fd, _sparse_areas) = self.get_region_info(
+                    index,
+                    Some(region_info.argsz - std::mem::size_of::<vfio_region_info>() as u32),
+                )?;
+                region_info = _region_info;
+                fd = _fd;
+                sparse_areas = _sparse_areas;
```

Comment:
I understand how you're trying to reuse the logic from `get_region_info()`, but IMO it would be nicer to have `get_region_info()` to handle everything transparently. This would simply the code here, even if that brings some slight duplication inside `get_region_info()`.

Reply:
Right. I tried your advice and it did turn out to be cleaner. I pushed a new commit to the PR to save some efforts on rebasing, and hope that's fine to you.